File: src/Sampling_Metropolis.f90
!====================================================================
! Basic Metropolis Rule for molecular systems. 
!====================================================================
module MetropolisRule
  use VarPrecision
  use CoordinateTypes, only: Perturbation, Addition, Deletion, VolChange
  use Template_AcceptRule, only: acceptrule
 
  type, public, extends(acceptrule) :: metropolis
    contains
       procedure, pass :: MakeDecision => Metropolis_MakeDecision
       procedure, pass :: MakeDecision2Box => Metropolis_MakeDecision2Box
!       procedure, pass :: Maintenance => Metropolis_Maintenance
!       procedure, pass :: ProcessIO => Metropolis_ProcessIO
  end type
!====================================================================
  contains
!====================================================================
  function Metropolis_MakeDecision(self, trialBox, E_Diff, disp, inProb, logProb, extraIn) result(accept)
    use Template_SimBox, only: SimBox
    use ErrorChecking, only: IsNan, IsInf
    use RandomGen, only: grnd
    use ParallelVar, only: nout
    implicit none
    class(metropolis), intent(inout) :: self
    class(simBox), intent(in) :: trialBox
    class(Perturbation), intent(in) :: disp(:)
    real(dp), intent(in), optional:: inProb, logProb
    real(dp), intent(in), optional:: extraIn
    real(dp), intent(in) :: E_Diff
    logical :: accept
    integer :: iDisp
    real(dp) :: biasE = 1E0_dp
    real(dp) :: chemPot, extraTerms, probTerm


    if(IsNan(E_Diff)) then
      write(0,*) "ERROR! Invalid energy has been passed to the sampling routine!"
      write(0,*) E_Diff
      error stop
    endif

    accept = .false.
    if(present(inProb)) then
      if(inProb <= 0E0_dp) then
        return
!      write(nout,*) "Probability:", inProb
!      error stop "CRITICAL ERROR! Probability passed to the Metropolis Sampling Function is zero or negative!"
      endif
    endif

    if(present(extraIn)) then
      extraTerms = extraIn
    else
      extraTerms = 0E0_dp
    endif

    if(present(inProb)) then
      probTerm = log(inProb)
    elseif(present(logProb)) then
      probTerm = logProb
    else
      write(0,*) "Coding Error! Probability has not been passed into Sampling "
      error stop
    endif
!    write(*,*) "Bias", E_diff
!    write(*,*) "    ", probTerm, extraTerms

    biasE = -trialBox%beta * E_Diff + probTerm + extraTerms
!    write(*,*) "    ",  biasE
!    write(*,*)
    if(biasE >= 0.0E0_dp) then
      accept = .true.
    elseif( biasE > log(grnd()) ) then
      accept = .true.
    endif

  end function
!====================================================================
  function Metropolis_MakeDecision2Box(self, trialBox1,  trialBox2, E_Diff1, E_Diff2, &
                                       disp1, disp2, inProb, &
                                       logProb, extraIn ) result(accept)
    use Template_SimBox, only: SimBox
    use RandomGen, only: grnd
    use ParallelVar, only: nout
    implicit none
    class(Metropolis), intent(inout) :: self
    class(SimBox), intent(in) :: trialBox1, trialBox2
    class(Perturbation), intent(in) :: disp1(:), disp2(:)
    real(dp), intent(in) :: E_Diff1, E_Diff2
    real(dp), intent(in), optional :: inProb, logProb, extraIn
    logical :: accept
    integer :: iDisp
    real(dp) :: biasE, chemPot, extraTerms, probTerm



    accept = .false.
    if(present(inProb)) then
      if(inProb <= 0E0_dp) then
        return
      endif
    endif

    if(present(extraIn)) then
      extraTerms = extraIn
    else
      extraTerms = 0E0_dp
    endif

    if(present(inProb)) then
      probTerm = log(inProb)
    elseif(present(logProb)) then
      probTerm = logProb
    else
      write(0,*) "Coding Error! Probability has not been passed into Sampling "
      error stop
    endif

!    write(*,*) "Bias", E_diff1, E_Diff2
!    write(*,*) "    ", probTerm, extraTerms

    biasE = -trialBox1%beta*E_Diff1 - trialBox2%beta*E_Diff2 + probTerm + extraTerms
!    write(*,*) "    ",  biasE
!    write(*,*)
    if(biasE > 0.0E0_dp) then
      accept = .true.
    elseif( biasE > log(grnd()) ) then
      accept = .true.
    endif

  end function

!====================================================================
end module
!====================================================================

File: src_python/Sampling_Metropolis.py
import math
import sys
import numpy as np
from src_python.VarPrecision import dp
from src_python.Template_AcceptRule import AcceptRule

# Simple implementations for missing functions
def IsNan(value):
    """Check if value is NaN"""
    return np.isnan(value)

np.random.random()

class Metropolis(AcceptRule):
    """Basic Metropolis Rule for molecular systems."""
    
    def __init__(self):
        super().__init__()
    
    def make_decision(self, trial_box, e_diff, disp, in_prob=None, log_prob=None, extra_in=None):
        """
        Make acceptance decision based on Metropolis criterion.
        
        Args:
            trial_box: SimBox instance
            e_diff: Energy difference (float)
            disp: List of Perturbation objects
            in_prob: Input probability (optional)
            log_prob: Log probability (optional)
            extra_in: Extra terms (optional)
            
        Returns:
            bool: Whether to accept the move
        """
        if IsNan(e_diff):
            print("ERROR! Invalid energy has been passed to the sampling routine!", file=sys.stderr)
            print(e_diff, file=sys.stderr)
            raise RuntimeError("Invalid energy in sampling routine")
        
        accept = False
        
        if in_prob is not None:
            if in_prob <= 0.0:
                return False
        
        if extra_in is not None:
            extra_terms = extra_in
        else:
            extra_terms = 0.0
        
        if in_prob is not None:
            prob_term = math.log(in_prob)
        elif log_prob is not None:
            prob_term = log_prob
        else:
            print("Coding Error! Probability has not been passed into Sampling", file=sys.stderr)
            raise RuntimeError("No probability provided")
        
        bias_e = -trial_box.beta * e_diff + prob_term + extra_terms
        
        if bias_e >= 0.0:
            accept = True
        elif bias_e > math.log(np.random.random()):
            accept = True
        
        return accept
    # ---------------------------------------------------------------------------
    def make_decision_2box(self, trial_box1, trial_box2, e_diff1, e_diff2, 
                          disp1, disp2, in_prob=None, log_prob=None, extra_in=None):
        """
        Make acceptance decision for two-box system.
        
        Args:
            trial_box1: First SimBox instance
            trial_box2: Second SimBox instance
            e_diff1: Energy difference for box 1
            e_diff2: Energy difference for box 2
            disp1: List of Perturbation objects for box 1
            disp2: List of Perturbation objects for box 2
            in_prob: Input probability (optional)
            log_prob: Log probability (optional)
            extra_in: Extra terms (optional)
            
        Returns:
            bool: Whether to accept the move
        """
        accept = False
        
        if in_prob is not None:
            if in_prob <= 0.0:
                return False
        
        if extra_in is not None:
            extra_terms = extra_in
        else:
            extra_terms = 0.0
        
        if in_prob is not None:
            prob_term = math.log(in_prob)
        elif log_prob is not None:
            prob_term = log_prob
        else:
            print("Coding Error! Probability has not been passed into Sampling", file=sys.stderr)
            raise RuntimeError("No probability provided")
        
        bias_e = (-trial_box1.beta * e_diff1 - 
                  trial_box2.beta * e_diff2 + 
                  prob_term + extra_terms)
        
        if bias_e > 0.0:
            accept = True
        elif bias_e > math.log(np.random.random()):
            accept = True
        
        return accept
